src/KMonad/Util/Keyboard/Mac.hs

```diff
@@ -0,0 +1,197 @@
+  , keycodeNames
+  )
+
+where
+
+
+import KMonad.Prelude
+
+import Foreign.Storable
+import KMonad.Util.Keyboard.Common
+import KMonad.Util.Keyboard.Keynames
+
+import qualified RIO.HashMap as M
+import qualified RIO.Text    as T
+import RIO.List.Partial ((!!))
+
+
+--------------------------------------------------------------------------------
+-- $keycode
+
+-- first Word32 represents the IOKit usage page
+-- second Word32 represent the IOKit usage
+newtype Keycode = Keycode { unKeycode :: (Word32, Word32) }
+  deriving (Eq, Ord, Hashable)
+
+instance Show Keycode where
+  show (Keycode (p, u)) = "(" ++ toHex p ++ ", " ++ toHex u ++ ")"
+
+-- https://stackoverflow.com/questions/1959715/how-to-print-integer-literals-in-binary-or-hex-in-haskell/55623656#55623656
+toHex :: Word32 -> String
+toHex = ("0x" ++) . convertToBase 16
+  where
+    convertToBase :: Word8 -> Word32 -> String
+    convertToBase b n
+      | n < 0              = '-' : convertToBase b (-n)
+      | n < fromIntegral b = [(['0'..'9'] ++ ['A' .. 'Z']) !! fromIntegral n]
+      | otherwise          = let (d, m) = n `divMod` fromIntegral b in convertToBase b d ++ convertToBase b m
+--------------------------------------------------------------------------------
+  -- $evtype
+
+-- | Mac has two types types of events: Press and Release.
+--
+-- The type is ordered in a way to line up with the integers our Mac API uses to
+-- encode for these events. I.e. 0=Press, 1=Release
+data EvType = MacPress | MacRelease
+  deriving (Eq, Show, Enum)
+
+-- | Prism between Mac and Haskell representation of event types
+_EvType :: Prism' Word64 EvType
+_EvType = prism' (fi . fromEnum) $ \i ->
+  ([MacPress, MacRelease] ^? ix (fi i))
+
+
+--------------------------------------------------------------------------------
+-- $raw
+
+-- | The RawEvent datatype
+--
+-- It contains a 'Word64' signifying whether the event was a
+-- Press (0) or Release (1), and two 'Word32's (uint32_t)
+-- signifying the Mac keycode:
+--   first: represents the IOKit usage page.
+--   second: represent the IOKit usage.
+data RawEvent = RawEvent
+  { _reVal  :: Word64  -- ^ press or release
+  , _reCode :: (Word32, Word32) -- ^ keycode
+  } deriving (Eq, Ord, Show)
+makeClassy ''RawEvent
+
+-- | This lets us send key events between Haskell and C.
+instance Storable RawEvent where
+  alignment _ = 4
+  sizeOf    _ = 16
+  peek ptr = do
+    s <- peekByteOff ptr 0
+    p <- peekByteOff ptr 8
+    u <- peekByteOff ptr 12
+    return $ RawEvent s (p, u)
+  poke ptr (RawEvent s (p, u)) = do
+    pokeByteOff ptr 0 s
+    pokeByteOff ptr 8 p
+    pokeByteOff ptr 12 u
+
+-- | Make a new raw event
+mkRaw :: EvType -> Keycode -> RawEvent
+mkRaw p c = RawEvent
+  { _reCode = unKeycode c
+  , _reVal  = p ^. re _EvType }
+
+
+--------------------------------------------------------------------------------
+-- $keynames
+
+-- | zip 2 lists only if they have the same length.
+--   Caution: only ever use this to generate static data.
```

Comment:
I think this is a bit misleading, as it suggests that the only thing that's happening is a length comparison and then a `zip`

Reply:
What would be a better description?

Reply:
Good question :)  Since this is not polymorphic, I guess we could write exactly what's going on; something about building up named keycode pairs, perhaps

Reply:
```
-- | Takes a usage page, a list of usages and a list of names, 
--    and returns a list of tuples of keycodes and their names.
--
-- Caution: Only use this for static data as this throws an error
-- if the lengths of lists differ.
```

I am not familiar with Haddock; if there is any mistake in markup, feel free to correct me.

Reply:
Looks good to me :+1: